File: leetcode-submissions/0002-add-two-numbers/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int rem = 0;
        ListNode temp;
        ListNode ans = new ListNode(0);
        temp = ans;
        while (l1 != null || l2 != null) {
            int sum;
            if (l1 == null) {
                sum = l2.val + rem;
                l2 = l2.next;
            } else if (l2 == null) {
                sum = l1.val + rem;
                l1 = l1.next; 
            } else {
                sum = l1.val + l2.val + rem;
                l2 = l2.next;
                l1 = l1.next;
            }

            if (sum < 10) {
                rem = 0;
            } else {
                rem = 1;
                sum = sum % 10;
            }
            ans.next = new ListNode(sum);
            ans = ans.next;
        }
        if (rem == 1) {
            ans.next = new ListNode(1);
        }
        return temp.next;
    }
}

File: leetcode-submissions/0002-add-two-numbers/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        # Create a dummy node
        # Itearate over both linked lists
        # Keep a carrier
        # Add the current values of both nodes if they exist, else add 0
        # Update the carrier if necessary
        # Move the pointers
        dummy = ListNode(0)
        p = dummy
        p1 = l1
        p2 = l2
        carrier = 0
        while p1 or p2:
            n1 = p1.val if p1 else 0
            n2 = p2.val if p2 else 0
            tot = n1 + n2 + carrier
            if tot >= 10:
                stay = tot % 10 
                carrier = tot // 10
                p.next = ListNode(stay)
                p = p.next
            else:
                carrier = 0
                p.next = ListNode(tot)
                p = p.next
            p1 = p1.next if p1 else None
            p2 = p2.next if p2 else None
        if carrier != 0:
            p.next = ListNode(carrier)

        return dummy.next
        
